### PR TITLE
Sorting

### PR DESCRIPTION
**Pull request by joshrdane**
**Sorting Options**
- Name (default) `--sort name`
- Most recently accessed date `--sort adate`
- Most recently created date `--sort cdate`
- Most recently modified date `--sort mdate`

**Notes**
- `alt-s` cycles the sorting type
- Setting the sort flag only sets the starting sorting method. 
- As of right now, history does not keep track of which sorting method was used last. 
- There is not currently a column displaying the dates associated with the sorting methods when sorting is done by date. 
- The sorting method is shown in the footer as `sort:<type>`.

### CODE REVIEW

**Comment by mgunyho** [comment]
Thanks! I'm busy this weekend but I'll take a look early next week.

As a quick note, `--sort X` is kind of inconsistent with other similar "cyclable" options (like e.g. case sensitive mode), but I think this is definitely the right way to do it. I'm not sure why I did it the other way with the others tbh, I should think about that again.

**Review by mgunyho — CHANGES_REQUESTED** [review]
Sorry for the delay! This looks pretty good, the review comments are mostly cosmetic. Some general comments:

- After merging #62, `app_state.settings` has to be changed to `app_state.settings()` everywhere, sorry about that.
- The search state is not correctly preserved when changing the sort mode, see gif below. Try to add a setter for the sort mode in `app_state` and use `with_cursor_fixed_at_current_item` like `set_filter_search` does (added in #62). Please add some simple tests to `app_state` to make sure that it works correctly.
- The info bar is getting quite crowded. For example in a folder with >100 files, having a terminal window with 70 characters (surprisingly common occurrence on my tiny laptop screen) and sorting by e.g. "accessed", the search overlaps with the info message even with a single character. I can try to squeeze it a bit after this is merged, but it has to be definitely done before the next release.
- I'm not entirely happy that we don't show the thing we're sorting by, but we can leave that for later (I think it would need a column header).


[sorting-search-not-updated2.webm](https://user-images.githubusercontent.com/20118130/191841296-25057d55-eec2-4833-a836-bfef15a161b2.webm)

**Comment by joshrdane** [comment]
Yeah, the column header is something that we can put in there and that would also free up the area I allocated for sorting in the footer. I will take some time this weekend to address the issues in the comments.

File: src/app_state.rs [hunk 1]
```diff
@@ -376,12 +396,25 @@ impl TereAppState {
...
+                        AttributeSortMode::Name => {
+                            // both are dirs or files, compare by name.
+                            // partial_cmp for strings always returns Some, so unwrap is ok here
+                            a.file_name_checked()
+                                .to_lowercase()
+                                .partial_cmp(&b.file_name_checked().to_lowercase())
+                                .unwrap()
+                        }
+                        AttributeSortMode::AccessedDate => {
+                            b.accessed().partial_cmp(&a.accessed()).unwrap()
```
[the remark this answers is not in the document]
Comment (joshrdane) [reply src/app_state.rs:409]: I agree. I also was not 100% on whether we should be sorting ASC or DSC or if we wanted that to be another feature added.

**Review by joshrdane — COMMENTED** [review]
I implemented everything except tests for the sorting mode based on date, I am having trouble getting the tests to populate times for the CustomDirEntries.

**Comment by mgunyho** [comment]
Merging the master made the diff a bit harder to interpret, I'll do a rebase on to develop.

**Comment by mgunyho** [comment]
Oops, that turned out to be a bit trickier than I thought. I cleaned up the version history a bit, split some things off into separate commits and removed some unnecessary ones (like importing intrinsics::unreachable and then removing the import). You will need to do `git fetch` and then `git reset --hard origin/sorting` to have your local version match up with this one, just make sure that you don't lose any work. I won't have time to do the review anymore today, but I'll get back to it soon.

**Comment by mgunyho** [comment]
Ah, one more thing that is missing: the new feature should be mentioned in the changelog, and also the keyboard shortcut and CLI option should be documented in the readme (there is a test to check that all default shortcuts are listed).

**Review by mgunyho — COMMENTED** [review]
So this looks pretty good. The following things are left:
- Should we leave out the option to sort by `accessed`? I'm a bit inclined towards yes.
- The README should mention the `--sort` CLI option, as well as the default keyboard shortcut and the `ChangeSortMode` action
- The CHANGELOG should mention this feature under v 1.3.0

I can make the changes to the readme and the changelog if you want.

**Comment by mgunyho** [comment]
Oh yeah and also I tried to create some unit tests for this, but you're right, modifying the timestamp of a `Metadata` is not really possible, and `update_ls_output_buf` is hardcoded to read the current folder, which you can't really change inside a test... so let's just rely on manual testing for now.

File: src/app_state.rs [hunk 2]
```diff
@@ -122,6 +123,27 @@ impl CustomDirEntry {
             None => false,
         }
     }
+
+    pub fn accessed(&self) -> SystemTime {
+        match &self.metadata {
+            Some(m) => m.accessed().unwrap_or(SystemTime::UNIX_EPOCH),
+            None => SystemTime::UNIX_EPOCH,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mgunyho) [reply src/app_state.rs:132]: Hmm, according to [this documentation](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/ns-fileapi-win32_file_attribute_data), the accessed time on Windows doesn't contain the time of day, just the date. So it's quite inconsistent across platforms, and not a very meaningful thing to sort by. Now I'm definitely convinced that we should remove it.

[the remark this answers is not in the document]
Comment (joshrdane) [reply src/app_state.rs:132]: I agree. I will remove it on my next push.

File: src/app_state.rs [hunk 3]
```diff
@@ -518,6 +554,14 @@ impl TereAppState {
         self.advance_search(""); // hacky, see the comment above in set_case_sensitive
     }
 
+    pub fn set_sort_mode(&mut self, sort_mode: SortMode) {
+        self.with_cursor_fixed_at_current_item(|self_| {
+            self_._settings.sort_mode = sort_mode;
+            self_.update_ls_output_buf().ok();
+            self_.advance_search("");
```
[the remark this answers is not in the document]
Comment (mgunyho) [reply src/app_state.rs:560]: I tried to add a separate method for sorting yesterday, but that wasn't really easy either, updating the `ls_output_buf` is quite a tangled mess. I should clean it up a bit, but that's for later.

Changing this line from `self_.advance_search("")` to `self_.update_search_matches()` seems to work and avoids doing some extra stuff. Interesgingly, it doesn't work for `set_case_sensitive` and `set_gap_search` above. But anyway, I think `update_search_matches` is good here.

File: src/settings.rs [hunk 4]
```diff
@@ -58,6 +58,31 @@ impl fmt::Display for GapSearchMode {
...
+impl Default for SortMode {
+    fn default() -> Self {
+        Self::Name
+    }
+}
+
+impl fmt::Display for SortMode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let text = match self {
+            SortMode::Name         => "name",
```
Comment (mgunyho) [inline src/settings.rs:77]:
```suggestion
            SortMode::Name => "name",
```

**Comment by mgunyho** [comment]
I went ahead and implemented these fixes myself, I hope you don't mind.

I'll do another rebase and then merge this.

**Comment by mgunyho** [comment]
Thanks a lot!

**Pull request merged**